File: kanban.rb
# kanban.rb
require 'json'
require 'date'

class Status
  TODO = "To Do"
  IN_PROGRESS = "In Progress"
  DONE = "Done"
end

class Priority
  LOW = "Низкий"
  MEDIUM = "Средний"
  HIGH = "Высокий"
end

class Task
  attr_accessor :id, :title, :description, :status, :priority, :created_at

  def initialize(id, title, description, status = Status::TODO, priority = Priority::MEDIUM, created_at = DateTime.now.iso8601)
    @id = id
    @title = title
    @description = description
    @status = status
    @priority = priority
    @created_at = created_at
  end

  def to_h
    {
      id: @id,
      title: @title,
      description: @description,
      status: @status,
      priority: @priority,
      created_at: @created_at
    }
  end

  def self.from_h(hash)
    Task.new(hash[:id], hash[:title], hash[:description], hash[:status], hash[:priority], hash[:created_at])
  end
end

class KanbanBoard
  attr_reader :tasks

  def initialize
    @tasks = []
    @next_id = 1
  end

  def add_task(title, description, priority = Priority::MEDIUM)
    task = Task.new(@next_id, title, description, Status::TODO, priority)
    @tasks << task
    @next_id += 1
    task
  end

  def find_task(id)
    @tasks.find { |t| t.id == id }
  end

  def move_task(id, new_status)
    task = find_task(id)
    return false unless task
    task.status = new_status
    true
  end

  def delete_task(id)
    task = find_task(id)
    return false unless task
    @tasks.delete(task)
    true
  end

  def filter_by_status(status)
    @tasks.select { |t| t.status == status }
  end

  def filter_by_priority(priority)
    @tasks.select { |t| t.priority == priority }
  end

  def save_to_file(filename = "kanban_data.json")
    File.write(filename, JSON.pretty_generate(@tasks.map(&:to_h)))
  end

  def load_from_file(filename = "kanban_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @tasks.clear
    data.each do |item|
      task = Task.from_h(item)
      @tasks << task
      @next_id = task.id + 1 if task.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end
end

def display_task(task)
  emoji = { "To Do" => "📌", "In Progress" => "🔄", "Done" => "✅" }
  puts "#{emoji[task.status]} ##{task.id} - #{task.title} (#{task.status})"
  puts "   Описание: #{task.description}"
  puts "   Приоритет: #{task.priority}, Создана: #{task.created_at}"
end

def main
  board = KanbanBoard.new
  board.load_from_file

  loop do
    puts "\n===== КАНБАН-ДОСКА (Ruby) ====="
    puts "1. Показать все задачи"
    puts "2. Добавить задачу"
    puts "3. Переместить задачу (изменить статус)"
    puts "4. Удалить задачу"
    puts "5. Сохранить в файл"
    puts "6. Загрузить из файла"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      if board.tasks.empty?
        puts "Задач нет."
      else
        board.tasks.each { |t| display_task(t) }
      end
    when "2"
      print "Введите название: "
      title = gets.chomp
      next if title.empty?
      print "Введите описание: "
      desc = gets.chomp
      puts "Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий"
      prio = gets.chomp
      priority = case prio
                 when "1" then Priority::LOW
                 when "3" then Priority::HIGH
                 else Priority::MEDIUM
                 end
      task = board.add_task(title, desc, priority)
      puts "Задача добавлена с ID #{task.id}"
    when "3"
      print "Введите ID задачи: "
      id = gets.chomp.to_i
      task = board.find_task(id)
      unless task
        puts "Задача не найдена."
        next
      end
      puts "Текущий статус: #{task.status}"
      puts "Выберите новый статус: 1-To Do, 2-In Progress, 3-Done"
      new_choice = gets.chomp
      new_status = case new_choice
                   when "1" then Status::TODO
                   when "2" then Status::IN_PROGRESS
                   when "3" then Status::DONE
                   else nil
                   end
      unless new_status
        puts "Неверный выбор."
        next
      end
      if board.move_task(id, new_status)
        puts "Статус обновлён."
      else
        puts "Ошибка."
      end
    when "4"
      print "Введите ID задачи для удаления: "
      id = gets.chomp.to_i
      if board.delete_task(id)
        puts "Задача удалена."
      else
        puts "Задача не найдена."
      end
    when "5"
      board.save_to_file
      puts "Сохранено."
    when "6"
      board.load_from_file
      puts "Загружено."
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0

File: kanban.py
// kanban.py
import json
import sys
from dataclasses import dataclass, asdict
from enum import Enum, auto
from typing import List, Optional
from datetime import datetime

class Status(Enum):
    TODO = "To Do"
    IN_PROGRESS = "In Progress"
    DONE = "Done"

class Priority(Enum):
    LOW = "Низкий"
    MEDIUM = "Средний"
    HIGH = "Высокий"

@dataclass
class Task:
    id: int
    title: str
    description: str
    status: Status
    priority: Priority
    created_at: str

class KanbanBoard:
    def __init__(self):
        self.tasks: List[Task] = []
        self.next_id = 1

    def add_task(self, title: str, desc: str, priority: Priority) -> Task:
        task = Task(
            id=self.next_id,
            title=title,
            description=desc,
            status=Status.TODO,
            priority=priority,
            created_at=datetime.now().isoformat()
        )
        self.tasks.append(task)
        self.next_id += 1
        return task

    def get_task(self, task_id: int) -> Optional[Task]:
        return next((t for t in self.tasks if t.id == task_id), None)

    def move_task(self, task_id: int, new_status: Status) -> bool:
        task = self.get_task(task_id)
        if task:
            task.status = new_status
            return True
        return False

    def delete_task(self, task_id: int) -> bool:
        task = self.get_task(task_id)
        if task:
            self.tasks.remove(task)
            return True
        return False

    def filter_by_status(self, status: Status) -> List[Task]:
        return [t for t in self.tasks if t.status == status]

    def filter_by_priority(self, priority: Priority) -> List[Task]:
        return [t for t in self.tasks if t.priority == priority]

    def save_to_file(self, filename: str = "kanban_data.json") -> None:
        with open(filename, "w", encoding="utf-8") as f:
            json.dump([asdict(t) for t in self.tasks], f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "kanban_data.json") -> None:
        try:
            with open(filename, "r", encoding="utf-8") as f:
                data = json.load(f)
                self.tasks.clear()
                for item in data:
                    task = Task(
                        id=item["id"],
                        title=item["title"],
                        description=item["description"],
                        status=Status(item["status"]),
                        priority=Priority(item["priority"]),
                        created_at=item["created_at"]
                    )
                    self.tasks.append(task)
                    if task.id >= self.next_id:
                        self.next_id = task.id + 1
        except FileNotFoundError:
            print("Файл не найден, создана пустая доска.")

def print_task(task: Task, color: bool = True) -> None:
    status_emoji = {"To Do": "📌", "In Progress": "🔄", "Done": "✅"}
    print(f"{status_emoji[task.status.value]} #{task.id} - {task.title} ({task.status.value})")
    print(f"   Описание: {task.description}")
    print(f"   Приоритет: {task.priority.value}, Создана: {task.created_at}")

def main():
    board = KanbanBoard()
    board.load_from_file()

    while True:
        print("\n===== КАНБАН-ДОСКА (Python) =====")
        print("1. Показать все задачи")
        print("2. Добавить задачу")
        print("3. Переместить задачу (изменить статус)")
        print("4. Удалить задачу")
        print("5. Сохранить в файл")
        print("6. Загрузить из файла")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            if not board.tasks:
                print("Задач нет.")
                continue
            for t in board.tasks:
                print_task(t)
        elif choice == "2":
            title = input("Введите название: ").strip()
            if not title:
                print("Название не может быть пустым.")
                continue
            desc = input("Введите описание: ").strip()
            print("Выберите приоритет: 1-Низкий, 2-Средний, 3-Высокий")
            prio_choice = input("> ").strip()
            if prio_choice == "1":
                priority = Priority.LOW
            elif prio_choice == "2":
                priority = Priority.MEDIUM
            elif prio_choice == "3":
                priority = Priority.HIGH
            else:
                priority = Priority.MEDIUM
            task = board.add_task(title, desc, priority)
            print(f"Задача добавлена с ID {task.id}")
        elif choice == "3":
            try:
                task_id = int(input("Введите ID задачи: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            task = board.get_task(task_id)
            if not task:
                print("Задача не найдена.")
                continue
            print(f"Текущий статус: {task.status.value}")
            print("Выберите новый статус: 1-To Do, 2-In Progress, 3-Done")
            new_choice = input("> ").strip()
            if new_choice == "1":
                new_status = Status.TODO
            elif new_choice == "2":
                new_status = Status.IN_PROGRESS
            elif new_choice == "3":
                new_status = Status.DONE
            else:
                print("Неверный выбор.")
                continue
            if board.move_task(task_id, new_status):
                print("Статус обновлён.")
            else:
                print("Ошибка.")
        elif choice == "4":
            try:
                task_id = int(input("Введите ID задачи для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if board.delete_task(task_id):
                print("Задача удалена.")
            else:
                print("Задача не найдена.")
        elif choice == "5":
            board.save_to_file()
            print("Сохранено.")
        elif choice == "6":
            board.load_from_file()
            print("Загружено.")
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()
